sbt/src/test/scala/testpkg/ServerSpec.scala

```diff
@@ -216,13 +223,24 @@ case class TestServer(baseDirectory: File)(implicit ec: ExecutionContext) {
 
   @tailrec
   final def waitForString(num: Int)(f: String => Boolean): Boolean = {
-    if (num < 0) false
-    else
-      readFrame match {
-        case Some(x) if f(x) => true
-        case _ =>
-          waitForString(num - 1)(f)
+    if (num < 0) { throw new Exception("Retries are over.") } else {
+      val result = Future { readFrame }(ec)
+      import scala.concurrent.duration._
+      Try {
+        Await.result(
+          result.map(
+            _ match {
+              case Some(x) if f(x) => true
+              case _               => false
+            }
+          ),
+          num seconds
```

Comment:
Did you mean to wait less and less each iteration? In my simplification, I just used a constant timeout, but I'm not sure if this was intentional enough. Also, I prefer `num.seconds` in this kind of instance.

Reply:
+1 on `num.seconds`

Reply:
yup, that's better in this case

Reply:
no more needed with the `CountDownLatch`